src/poly/cse.rs

```diff
@@ -0,0 +1,77 @@
+use rand_chacha::{rand_core::SeedableRng, ChaCha20Rng};
+
+use crate::{field::Field, sbpir::query::Queriable};
+
+use std::{fmt::Debug, hash::Hash};
+
+use super::{Expr, HashResult, VarAssignments};
+
+/// Common Subexpression Elimination - takes a collection of expr
+pub fn cse<F: Field + Hash, V: Debug + Clone + Eq + Hash>(
+    exprs: Vec<Expr<F, V, ()>>,
+    queriables: &Vec<Queriable<F>>,
+) -> Vec<Expr<F, V, HashResult>> {
+    // generate random point in the field set
+    let _assignments = generate_random_assignment(queriables);
+
+    // hash all the functions using the point for the assignments
+
+    // check if the hash is already in the hash table
+    // if it is, then replace it by a reference to the other one
+
+    // return the expressions hashed and optimized
+
+    todo!()
+}
+
+fn generate_random_assignment<F: Field + Hash>(
+    queriables: &Vec<Queriable<F>>,
+) -> VarAssignments<F, String> {
+    let mut rng = ChaCha20Rng::seed_from_u64(0);
+
+    let mut assignments = VarAssignments::new();
+
+    for queriable in queriables {
+        let value = F::random(&mut rng);
```

Comment:
If the queriable is Forward, it should produce two values, one with false and another with true in `.1`